### ISSUE TITLE
concurrency.Session fails if there is reconnection to server

### ISSUE THREAD

**Issue by nekto0n** [coreos/etcd#5571]
Hi!
I am experimenting with etcdv3 (built from master). I tried to use `concurrency.Session` primitive and seems that it can stuck if client connection to server fails and then reestablishes. I came up with code like this:
```
package housekeeping

import (
	"github.com/coreos/etcd/clientv3"
	"github.com/coreos/etcd/clientv3/concurrency"
	"github.com/golang/glog"
	"golang.org/x/net/context"
	"time"
)

var nodeValue string = "Lock"

type CancellableFunc func(context.Context)

func notFound(key string) clientv3.Cmp {
	return clientv3.Compare(clientv3.ModRevision(key), "=", 0)
}

func LockAndRun(ctx context.Context, c *clientv3.Client, name string, runnable CancellableFunc) {
	glog.Infof("Started locking %s", name)
	for {
	session:
		cancelCtx, cancelFunc := context.WithCancel(ctx)
		s, err := concurrency.NewSession(c)
		if err == nil {
			glog.Info("Created session with lease ID: ", s.Lease())
			for {
				glog.V(2).Info("Iterating")
				txnResp, err := c.KV.Txn(ctx).If(
					notFound(name),
				).Then(
					clientv3.OpPut(name, nodeValue, clientv3.WithLease(s.Lease())),
				).Commit()
				if err != nil {
					if ctx.Err() != nil {
						// We have been cancelled
						return
					}
					// We might have disconnected and need to start a new session
					glog.Error("Failed to create lock file: ", err.Error())
					s.Orphan()
					time.Sleep(5 * time.Second) // Avoid busy looping
					goto session
				} else if !txnResp.Succeeded {
					glog.V(2).Info("Transaction failed, will retry...")
					time.Sleep(5 * time.Second)
				} else {
					break
				}
			}
			// Now we have created file with lease, assuming lock is held
			glog.Infof("Assuming control over '%s', starting runnable", name)
			go runnable(cancelCtx)
			select {
			case <-ctx.Done(): // If initial context is Done - stop all activities
				cancelFunc()
				return
			case <-s.Done(): // If session is Done, just start all over again
				cancelFunc()
				break
			}
			glog.Infof("Stopped runnable, will retry")
			continue
		} else if ctx.Err() != nil {
			glog.Info("Context cancelled ", ctx.Err().Error())
		} else {
			glog.Errorf("Failed to establish session: %s", err.Error())
			time.Sleep(5 * time.Second)
		}
	}
}
```

If I run it and restart etcd in another terminal, I get something like this
```
I0606 18:34:51.588098   30070 leader.go:23] Started locking /_leaders/housekeeper
I0606 18:34:51.588793   30070 leader.go:29] Created session with lease ID: 4596579991049219091
I0606 18:34:51.588806   30070 leader.go:31] Iterating
I0606 18:34:51.592221   30070 leader.go:55] Assuming control over '/_leaders/housekeeper', starting runnable
# At this point we are leading and runnable is running
I0606 18:36:17.425689   30070 leader.go:65] Stopped runnable, will retry
# etcd is shut down
E0606 18:36:17.425809   30070 leader.go:70] Failed to establish session: grpc: timed out trying to connect
E0606 18:36:22.426144   30070 leader.go:70] Failed to establish session: grpc: timed out trying to connect
# etcd is started
I0606 18:36:30.240222   30070 leader.go:29] Created session with lease ID: 4596579991076462850
I0606 18:36:30.240238   30070 leader.go:31] Iterating
I0606 18:36:30.241878   30070 leader.go:48] Transaction failed, will retry...
I0606 18:36:35.242068   30070 leader.go:31] Iterating
I0606 18:36:35.242682   30070 leader.go:48] Transaction failed, will retry...
I0606 18:36:40.242878   30070 leader.go:31] Iterating
I0606 18:36:40.243489   30070 leader.go:48] Transaction failed, will retry...
I0606 18:36:45.243656   30070 leader.go:31] Iterating
I0606 18:36:45.244288   30070 leader.go:48] Transaction failed, will retry...
I0606 18:36:50.244409   30070 leader.go:31] Iterating
I0606 18:36:50.245018   30070 leader.go:48] Transaction failed, will retry...
I0606 18:36:55.245211   30070 leader.go:31] Iterating
I0606 18:36:55.245900   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:00.246091   30070 leader.go:31] Iterating
I0606 18:37:00.246916   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:05.247146   30070 leader.go:31] Iterating
I0606 18:37:05.247941   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:10.248086   30070 leader.go:31] Iterating
I0606 18:37:10.248686   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:15.248883   30070 leader.go:31] Iterating
I0606 18:37:15.249545   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:20.249891   30070 leader.go:31] Iterating
I0606 18:37:20.250546   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:25.250735   30070 leader.go:31] Iterating
I0606 18:37:25.251353   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:30.251481   30070 leader.go:31] Iterating
I0606 18:37:30.252107   30070 leader.go:48] Transaction failed, will retry...
I0606 18:37:35.252273   30070 leader.go:31] Iterating
# At this point my code will call s.Orphan() and get stuck
E0606 18:37:35.252894   30070 leader.go:43] Failed to create lock file: etcdserver: requested lease not found
```

Any help will be appreciated.

**Comment by heyitsanthony** [comment]
@nekto0n I can't seem to reproduce this failure. There's not much in Orphan that can lock up (I read through the cancelation path and it looks OK). Can you `ctrl+|` when it hangs and post the trace here?

**Comment by nekto0n** [comment]
Yeah, I failed to reproduce this behavior today. But nevertheless I see strange things :)

Part 1
====
  * Run this http://pastebin.com/QZy5rNpv with `-logtostderr -v2` and etcd running
  * Restart `etcd` after we assume that we have control over key
  * See, that session is alive (i.e. we see `Tick` message repeating)

Here is `SIGQUIT` output: http://pastebin.com/36Mr5sGL. Even if etcd never comes back `session.Done()` never closes.

Part 2
====
  * Fix example by setting `DialTimeot` in `clientv3.Config` to 1 second and run again. 
  * Restart `etcd`, see that session is closed correctly, runnable stopped and new session created.
  * Wait for a minute for previous session lease to expire.
  * See that runnable is started again.
  * Restart `etcd` again - session is assumed alive, runnable keeps looping.

Here is `SIGQUIT` output for this case: http://pastebin.com/r9Xb7nm0

**Comment by heyitsanthony** [comment]
@nekto0n, 
(1) is because the lease code will keep retrying to keepalive until it gets some response from etcd. I think this can be safely fixed in the client by having a timeout that closes the keepalive channel if the delay for a response exceeds the lease duration. 
(2) I don't understand what's the bug with this one. The loop creates a new session on retry if the old one expired-- why shouldn't `runnable` start again?

**Comment by nekto0n** [comment]
In case (2) session is not closed and recreated after etcd being restarted second time, runnable does not restart second time - it just keeps running. Sorry if my description didn't make it clear.

**Comment by heyitsanthony** [comment]
@nekto0n I tried the code with #5603 and it seems OK. I launched a 3 member cluster and stopped the etcd on 127.0.0.1:2379 to cut off the keep-alives and expire the lease. The code ticks until expiration and finishes runnable. I repeated this process and runnable goes up/down with the lease each time.

Let me know if it's still broken in case I missed something when trying to repro the issue. Thanks!
